File: Domain/Specialty.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Domain
{
    public class Specialty
    {
        public Guid Id { get; set; }
        public bool Visible { get; set; }
        public bool Approved { get; set; }
        public Institution Institution { get; set; }
        public SpecialtyCore SpecialtyCore { get; set; }
        public string Description { get; set; }
        public Degree Degree { get; set; }
        public decimal TuitionUAH { get; set; }
        public bool FreeEducation { get; set; }
        public double AcceptanceRate { get; set; }
        public double GraduationRate { get; set; }
        public double GraduateEmploymentRate { get; set; }
        public int UndergraduatesEnrolled { get; set; }
        public int EctsCredits { get; set; }
        public int StartYear { get; set; }
        public int EndYear { get; set; }
        public ICollection<StudyForm> StudyForms { get; set; } = new List<StudyForm>();
        public ICollection<Language> Languages { get; set; } = new List<Language>();
        public ICollection<Skill> Skills { get; set; } = new List<Skill>();
        public ICollection<Component> Components { get; set; } = new List<Component>();
    }
}

File: client-app/src/app/models/specialty.ts
import { EducationalComponent } from "./educationalComponent";

export interface Specialty {
    id: string;
    approved: boolean;
    visible: boolean;
    institutionId: string;
    description: string;
    localSpecialtyCode: string;
    degreeId: number;
    tuitionUSD: number;
    freeEducation: boolean;
    acceptanceRate: number;
    graduationRate: number;
    graduateEmploymentRate: number;
    undergraduatesEnrolled: number;
    ectsCredits: number;
    startYear: number;
    endYear: number;
    studyFormIds: number[];
    languageIds: string[];
    skillIds: number[];
    componentDTOs: EducationalComponent[];
}

export class Specialty implements Specialty {
    constructor(init?: SpecialtyFormValues | Specialty) {
        Object.assign(this, init);
    }
}

export class SpecialtyFormValues {
    id?: string = undefined;
    institutionId?: string = undefined;
    visible: boolean = false;
    localSpecialtyCode: string = '';
    description: string = '';
    degreeId: number = 1;
    tuitionUSD: number = 0;
    freeEducation: boolean = false;
    acceptanceRate: number = 0;
    graduationRate: number = 0;
    graduateEmploymentRate: number = 0;
    undergraduatesEnrolled: number = 0;
    ectsCredits: number = 0;
    startYear: number = 0;
    endYear: number = 0;
    studyFormIds: number[] = [];
    languageIds: string[] = [];
    skillIds: number[] = [];
    componentDTOs: EducationalComponent[] = [];

    constructor(specialty?: SpecialtyFormValues | Specialty) {
        if (specialty) {
            this.id = specialty.id;
            this.institutionId = specialty.institutionId;
            this.localSpecialtyCode = specialty.localSpecialtyCode;
            this.visible = specialty.visible;
            this.description = specialty.description;
            this.degreeId = specialty.degreeId;
            this.tuitionUSD = specialty.tuitionUSD;
            this.freeEducation = specialty.freeEducation;
            this.acceptanceRate = specialty.acceptanceRate;
            this.graduationRate = specialty.graduationRate;
            this.graduateEmploymentRate = specialty.graduateEmploymentRate;
            this.undergraduatesEnrolled = specialty.undergraduatesEnrolled;
            this.ectsCredits = specialty.ectsCredits;
            this.startYear = specialty.startYear;
            this.endYear = specialty.endYear;
            this.studyFormIds = specialty.studyFormIds;
            this.languageIds = specialty.languageIds;
            this.skillIds = specialty.skillIds;
            this.componentDTOs = specialty.componentDTOs;
        }
    }
}